cli/util/uploader.go

```diff
@@ -382,59 +385,61 @@ func updateIndexFiles(ctx *cli.Context, p *pool.Pool, containerID cid.ID, accoun
 					}
 					return
 				}
-				offset := (uint(blockIndex) % indexFileSize) * oidSize
-				id.Encode(buffer[offset:])
+				indexFilePosition := uint(blockIndex) % indexFileSize
+				if _, exists := processedIndices.LoadOrStore(int(indexFilePosition), blockIndex); !exists {
+					id.Encode(buffer[indexFilePosition*oidSize:])
+				}
 				oidFetcherToProcessor <- struct{}{}
 			}
 		}()
 	}
 
+	// Helper to wait for all objects to be processed
+	waitForCompletion := func(expectedCount int, errCh <-chan error, completionCh <-chan struct{}) error {
+		completed := 0
+		for completed < expectedCount {
+			select {
+			case err := <-errCh:
+				return err
+			case <-completionCh:
+				completed++
+			}
+		}
+		return nil
+	}
+
+	// Main processing loop for each index file
 	for i := existingIndexCount; i < expectedIndexCount; i++ {
 		startIndex := i * indexFileSize
 		endIndex := startIndex + indexFileSize
-		go func() {
-			for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
-				remaining := int(endIndex) - j
-				end := j + min(searchBatchSize, remaining)
-
-				prm = client.PrmObjectSearch{}
-				filters = object.NewSearchFilters()
-				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
-				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLT)
-				prm.SetFilters(filters)
-				var objIDs []oid.ID
-				err := retry(func() error {
-					var errSearchIndex error
-					objIDs, errSearchIndex = neofs.ObjectSearch(ctx.Context, p, account.PrivateKey(), containerID.String(), prm)
-					return errSearchIndex
-				})
-
+		objIDs, err := searchObj(ctx.Context, p, containerID, account, blockAttributeKey, startIndex, endIndex, maxParallelSearches)
+		if err != nil {
+			return err
+		}
+		for _, id := range objIDs {
+			oidCh <- id
+		}
+		if err = waitForCompletion(len(objIDs), errCh, oidFetcherToProcessor); err != nil {
+			return err
+		}
+		// Check if there are any empty oid in the created index file.
+		// This could happen if searchObj returned not all objects, or we have duplicates.
+		// In this case, we need to retry the search.
+		oidNum := 0
+		for idx := range indexFileSize {
+			if _, exists := processedIndices.Load(int(idx)); !exists {
+				objIDs, err = searchObj(ctx.Context, p, containerID, account, blockAttributeKey, i*indexFileSize+idx, i*indexFileSize+idx+1, 1)
 				if err != nil {
-					select {
-					case errCh <- fmt.Errorf("failed to search for objects from %d to %d for index file %d: %w", j, end, i, err):
-					default:
-					}
-					return
+					return err
 				}
-
+				oidNum += len(objIDs)
 				for _, id := range objIDs {
 					oidCh <- id
```

Comment:
The current version won't save from  https://github.com/nspcc-dev/neofs-node/issues/2988 - we will have an empty oid. Example:
`failed to update index files after upload: empty oid found in index file 0 at position 91 (block index 91)`

But if NeoFS don't mess up with attributes -everything will be okay.

another option:

```
-               oidNum := 0
                for idx := range indexFileSize {
                        if _, exists := processedIndices.Load(int(idx)); !exists {
                                objIDs, err = searchObj(ctx.Context, p, containerID, account, blockAttributeKey, i*indexFileSize+idx, i*indexFileSize+idx+1, 1)
                                if err != nil {
                                        return err
                                }
-                               oidNum += len(objIDs)
-                               for _, id := range objIDs {
-                                       oidCh <- id
-                               }
+                               processedIndices.Store(int(idx), objIDs[0])
+                               objIDs[0].Encode(buffer[idx*oidSize:])
                        }
                }
-               if err = waitForCompletion(oidNum, errCh, oidFetcherToProcessor); err != nil {
-                       return err
-               }

```

directly put the first one found without getting attributes.